experimental/stats/metricregistry.go

```diff
@@ -0,0 +1,240 @@
+	// intended to attached to metrics emitted from a stats handler if
+	// configured.
+	OptionalLabels []string
+	// Default is whether this metric is on by default.
+	Default bool
+	// Type is the type of metric. This is set by the metric registry, and not
+	// intended to be set by a component registering a metric.
+	Type MetricType
+}
+
+// Int64CountHandle is a typed handle for a float count metric. This handle
+// is passed at the recording point in order to know which metric to record
+// on.
+type Int64CountHandle struct {
+	MetricDescriptor *MetricDescriptor
+}
+
+// Record records the int64 count value on the metrics recorder provided.
+func (h Int64CountHandle) Record(recorder MetricsRecorder, incr int64, labels ...string) {
+	recorder.RecordIntCount(h, incr, labels...)
+}
+
+// Float64CountHandle is a typed handle for a float count metric. This handle is
+// passed at the recording point in order to know which metric to record on.
+type Float64CountHandle struct {
+	MetricDescriptor *MetricDescriptor
+}
+
+// Record records the float64 count value on the metrics recorder provided.
+func (h Float64CountHandle) Record(recorder MetricsRecorder, incr float64, labels ...string) {
+	recorder.RecordFloatCount(h, incr, labels...)
+}
+
+// Int64HistoHandle is a typed handle for an int histogram metric. This handle
+// is passed at the recording point in order to know which metric to record on.
+type Int64HistoHandle struct {
+	MetricDescriptor *MetricDescriptor
+}
+
+// Record records the int64 histo value on the metrics recorder provided.
+func (h Int64HistoHandle) Record(recorder MetricsRecorder, incr int64, labels ...string) {
+	recorder.RecordIntHisto(h, incr, labels...)
+}
+
+// Float64HistoHandle is a typed handle for a float histogram metric. This
+// handle is passed at the recording point in order to know which metric to
+// record on.
+type Float64HistoHandle struct {
+	MetricDescriptor *MetricDescriptor
+}
+
+// Record records the float64 histo value on the metrics recorder provided.
+func (h Float64HistoHandle) Record(recorder MetricsRecorder, incr float64, labels ...string) {
+	recorder.RecordFloatHisto(h, incr, labels...)
+}
+
+// Int64GaugeHandle is a typed handle for an int gauge metric. This handle is
+// passed at the recording point in order to know which metric to record on.
+type Int64GaugeHandle struct {
+	MetricDescriptor *MetricDescriptor
+}
+
+// Record records the int64 histo value on the metrics recorder provided.
+func (h Int64GaugeHandle) Record(recorder MetricsRecorder, incr int64, labels ...string) {
+	recorder.RecordIntGauge(h, incr, labels...)
+}
+
+// registeredMetrics are the registered metric descriptor names.
+var registeredMetrics = make(map[stats.Metric]bool)
+
+// MetricsRegistry are all of the registered metrics.
+//
+// This is written to only at init time, and read only after that.
+var MetricsRegistry = make(map[stats.Metric]*MetricDescriptor)
```

Comment:
Why is this exported by the functions that access it aren't?  Shouldn't it be the other way around?

Reply:
Ah I completely forgot. I was using this in OpenTelemetry and wondering why we had a map in the first place since the name was in the values, but yes if we return a getter that gets a pointer for a metric name which will create a better API, we do need the map as we discussed previously. Thank you for pointing this out.